File: Data Structures & Algorithms/reverse-a-linked-list/submission-2.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

# By looping
# class Solution:
#     def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
#         curr = head
#         prev = None
#         while curr != None:
#             next = curr.next
#             curr.next = prev
#             prev = curr
#             curr = next

#         return prev


# recursion
class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if head == None or head.next == None:
            return head
        
        reversedList = self.reverseList(head.next)
        head.next.next = head
        head.next = None
        return reversedList

        
        
        

File: Data Structures & Algorithms/merge-two-sorted-linked-lists/submission-2.js
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     constructor(val = 0, next = null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */

class Solution {
    /**
     * @param {ListNode} list1
     * @param {ListNode} list2
     * @return {ListNode}
     */
    mergeTwoLists(list1, list2) {
        if (!list1) return list2;
        if (!list2) return list1;
        let head = null, prev = null;

        while(list1 && list2){
            if(list1.val< list2.val){
                if(!head) head = list1;
                if(prev) prev.next = list1;
                prev = list1;
                list1 = list1.next;
            }else{
                if(prev) prev.next = list2;
                if(!head) head = list2;
                prev = list2;
                list2 = list2.next;
            }
        }
        if(list1){
            prev.next = list1;
        }
        if(list2){
            prev.next = list2;
        }

        return head;
    }
}
